### ISSUE TITLE
self-referential assignment does not trigger a warning for unset local variables (UnboundLocalError vs NameError handling)

### ISSUE THREAD

**Issue by Pyrestone** [microsoft/pyright#10648]
## Environment data

-   Pylance version: 2025.6.1
-   OS and version: ubuntu 20.04
-   Python version (& distribution if applicable, e.g. Anaconda): 3.8 (native)
- VSCode version: 1.101.1

## Code Snippet

```python

def fun1(): # calling this Causes NameError: name 'x' is not defined
    a=x # this raises a warning because x is unknown

def fun2(): # calling this Causes UnboundLocalError: local variable 'x' referenced before assignment
    x=x # this should also raise a warning because x is unknown, but it doesn't

def fun3(): # calling this Causes UnboundLocalError: local variable 'x' referenced before assignment
    x+=1 # this should also trigger a warning

```

## Repro Steps

Look at it in vs code:

![Image](https://github.com/user-attachments/assets/e48cf652-0a8b-4719-b324-cc9248242457)

**Comment by Pyrestone** [comment]
thanks, adding this to my vs code settings produces the behavior I was expecting to see with typeCheckingMode: "off":
```yaml
"python.analysis.diagnosticSeverityOverrides": {
  "reportUnboundVariable": "warning"
}
```



Enabling even basic type checking throws so many errors on otherwise working code because frankly, libraries doing proper type annotation for multi-variant function parameters is just the wild west out there, so I keep that disabled.

Since this has nothing to do with type annotations, but is more akin to a syntax/logic error, IMO it should be a warning with type checking off, but the workaround is fine for me right now.

**Comment by erictraut** [comment]
@Pyrestone, as you discovered, the `'x' is not defined` and `'x' is unbound` checks are covered under separate diagnostic checks — `reportUndefinedVariable` and `reportUnboundVariable`. You can see that the Python interpreter also treats these conditions as distinct at runtime; it raises a `NameError` for the former and an `UnboundLocalError` for the latter.

We enable the `reportUndefinedVariable` check by default (even when the type checking mode is set to off) because this check produces few, if any, false positives. By contrast, the `reportUnboundVariable` check (and the related `reportPossiblyUnboundVariable`) relies on static code analysis and can produce false positives if type information is incomplete. 

For a full list of diagnostic checks available in pyright (and therefore in pylance) and their default settings based on the `typeCheckingMode`, refer to [this documentation](https://microsoft.github.io/pyright/#/configuration?id=diagnostic-settings-defaults). The diagnostic checks at the top of this table tend to be the "safer" ones that produce fewer false positives even if type information is incomplete.
